index.js

```diff
@@ -1,30 +1,194 @@
+                        back: {
+                            type: 'string',
+                            description:
+                                'Back content of a flashcard containing the answer to the question. Quotation marks should be escaped with a forward slash.',
+                        },
+                    },
+                },
+            },
+        },
+    },
+};
+const print_flashcard = {
+    name: 'print_flashcard',
+    description: 'Print a flashcard object',
+    parameters: {
+        type: 'object',
+        properties: {
+            front: {
+                type: 'string',
+                description:
+                    'Front content of a flashcard containing a question. Quotation marks should be escaped with a forward slash.',
+            },
+            back: {
+                type: 'string',
+                description:
+                    'Back content of a flashcard containing the answer to the question. Quotation marks should be escaped with a forward slash.',
+            },
+        },
+    },
+};
+
+const promptBuilder = ({
+    cardCount,
+    topics,
+    keywords,
+    example,
+    difficulty,
+}) => {
+    let prompt = `The user will provide you with a set of text which they would like to print this number of flashcards: ${cardCount}. `;
+    if (topics) {
+        prompt += `They have indicated the topic(s) they would like the flashcards to be about are the following: ${topics.join(
+            ', '
+        )}. `;
+    }
+    if (keywords) {
+        prompt += `They have indicated that the keyword(s) they would like some flashcards to be about are the following: ${keywords.join(
+            ', '
+        )}. `;
+    }
+    if (example) {
+        prompt += `They have provided an example question and answer to help guide the generation of cards. Here is the example front content of the card: "${example.front}" and here is the example back content of the card: "${example.back}". `;
+    }
+    if (difficulty) {
+        `The user has indicated that they would like the difficulty level of the cards to be of ${difficulty} difficulty. `;
+    }
+    return prompt;
+};
+
+const generate = async (inputText, prompt) => {
     const chatCompletion = await openai.createChatCompletion({
         model: 'gpt-3.5-turbo',
         messages: [
             {
                 role: 'system',
-                content: `You should generate a list of ${cardCount} flashcards with front and back content to quiz the user on the subject matter which they submit.`,
+                content: prompt,
             },
             {
                 role: 'user',
                 content: `${inputText}`,
             },
         ],
+        functions: [print_flashcard, print_flashcards],
+        max_tokens: 1000,
     });
-    console.log(chatCompletion.data);
-    console.log(chatCompletion.data.choices[0].message);
+    const { arguments } = chatCompletion.data.choices[0].message.function_call;
```

Comment:
why arguments?

Reply:
I am using the OpenAI API [function call](https://platform.openai.com/docs/api-reference/chat/create#chat/create-function_call) feature here to enforce a schema on the response they provide. The schema is defined in the `print_flashcards` fn here:

https://github.com/alex-808/flashcard-service/blob/cd647e23bce3d7fb39843bb757829daf806ae398/index.js#L11-L42

So the output that the API provides is actually what it thinks will be arguments to this `print_flashcards` function given the prompts and user input.

I found this was the best way to get a predictable response format from the API which otherwise would return arbitrary text. 

I understand how it would be confusing. Would you recommend adding a comment to make it clearer or maybe renaming the variable after destructuring?

Reply:
this is very interesting! this seems like a new openai feature. good learning for me as well.
yeah, this is solving almost exactly the problem that i told you about building a annotation based protobuf to define output schema. this seems to be the solution.